### ISSUE TITLE
Copy and paste doesn't work when content has children

### ISSUE THREAD

**Issue by wesleybl** [plone/volto#3849]
**Describe the bug**
If a Document has children, it is not possible to copy and paste this Document. We get the error:

```bash
2022-11-07 14:34:45 ERROR [Zope.SiteErrorLog:35][waitress-1] AttributeError: http://localhost:3000/@copy
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 167, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 376, in publish_module
  Module ZPublisher.WSGIPublisher, line 271, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 68, in call_object
  Module plone.rest.service, line 22, in __call__
  Module plone.restapi.services, line 19, in render
  Module plone.restapi.services.copymove.copymove, line 83, in reply
  Module OFS.CopySupport, line 317, in manage_pasteObjects
  Module OFS.CopySupport, line 226, in _pasteObjects
  Module zope.event, line 32, in notify
  Module zope.component.event, line 27, in dispatch
  Module zope.component._api, line 134, in subscribers
  Module zope.interface.registry, line 448, in subscribers
  Module zope.interface.adapter, line 899, in subscribers
  Module zope.component.event, line 36, in objectEventNotify
  Module zope.component._api, line 134, in subscribers
  Module zope.interface.registry, line 448, in subscribers
  Module zope.interface.adapter, line 899, in subscribers
  Module OFS.subscribers, line 136, in dispatchObjectCopiedEvent
  Module zope.container.contained, line 157, in dispatchToSublocations
  Module zope.component._api, line 138, in handle
  Module zope.interface.registry, line 448, in subscribers
  Module zope.interface.adapter, line 899, in subscribers
  Module plone.app.linkintegrity.handlers, line 116, in modifiedContent
  Module plone.restapi.blocks_linkintegrity, line 25, in retrieveLinks
AttributeError: REQUEST
```

**To Reproduce**

1. Create a document.
2. Inside that document, create another document.
3. Go to Contents, and try to copy and paste the parent document.

**Expected behavior**
Copy and paste works.

**Software (please complete the following information):**

- OS: Ubuntu 22
- Browser: Chrome
- Volto Version: 16.0.0-alpha.48
- Plone Version 6.0.0b3
- Plone REST API Version: 8.30.0

**Comment by tiberiuichim** [comment]
This is a plone.restapi issue.

**Comment by wesleybl** [comment]
@cekk can you please take a look here? It seems that it was after the PR https://github.com/plone/plone.restapi/pull/953 that this error started to occur.

**Comment by davisagli** [comment]
Using getRequest from zope.globalrequest instead of context.REQUEST should avoid this error (which happens if the object does not have a full acquisition chain to the Zope root)

**Comment by wesleybl** [comment]
@davisagli in fact, the classes that implement the interface don't even use request. For example:

https://github.com/plone/plone.restapi/blob/60293d86de9a89220052f434e92f01649b0b7029/src/plone/restapi/blocks_linkintegrity.py#L37-L39

So I wonder if the request could not be removed.

**Comment by tiberiuichim** [comment]
@wesleybl the request is useful as discriminators in the adaptors, as they allow overriding, per Plone site, the behavior, in case that's needed.

**Comment by wesleybl** [comment]
@tiberiuichim you mean through a layer? Adapter does not have the layer attribute. It is not possible to customize an adapter per site, through layer.

**Comment by tiberiuichim** [comment]
@wesleybl they don't have the layer, but the request will provide a marker layer interface. https://github.com/plone/bobtemplates.plone/blob/9bf1efef7c735ecec99d6e195aa5b10fa0e3dbfa/bobtemplates/plone/addon/src/+package.namespace+/+package.name+/profiles/default/browserlayer.xml.bob

**Comment by wesleybl** [comment]
@tiberiuichim are you proposing to customize the adapter and make an if in the custom adapter itself?

**Comment by tiberiuichim** [comment]
@wesleybl no... but I'm not sure I understand your last message correctly.

In any case, let me try explain:

- you were asking why it's a multiadapter with context+request
- I've explained that, by being a multi-adapter with the request as part of the "discriminators", it is possible to override, per Plone site, the behavior. It's a common practice
- you said "adapters don't have layers", indeed, but the request will `directlyProvides` layer marker interfaces
- so you can register an override not by doing it via a configure-overrides.zcml, but you can do it in normal code, by using a more specific layer interface instead of IBrowserRequest

**Comment by wesleybl** [comment]
> so you can register an override not by doing it via a configure-overrides.zcml, but you can do it in normal code, by using a more specific layer interface instead of IBrowserRequest

@tiberiuichim  in that case, you would not overwrite the original. You would have one more subscriber on the site with the more specific interface. More specific interfaces only overwrite components that have the layer abribute.

I don't know if that's what you were looking for, I don't even know if it's worth keeping the request because of that.

**Comment by tiberiuichim** [comment]
@wesleybl I see now the issue, the adapters are actually subscribers.
